File: sample-chat/src/main/java/com/quickblox/sample/chat/utils/qb/QbDialogHolder.java
package com.quickblox.sample.chat.utils.qb;

import com.quickblox.chat.model.QBChatDialog;
import com.quickblox.chat.model.QBChatMessage;
import com.quickblox.chat.model.QBDialogType;
import com.quickblox.users.model.QBUser;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class QbDialogHolder {

    private static QbDialogHolder instance;
    private Map<String, QBChatDialog> dialogsMap;

    public static synchronized QbDialogHolder getInstance() {
        if (instance == null) {
            instance = new QbDialogHolder();
        }
        return instance;
    }

    private QbDialogHolder() {
        dialogsMap = new TreeMap<>();
    }

    public Map<String, QBChatDialog> getDialogs() {
        return getSortedMap(dialogsMap);
    }

    public QBChatDialog getChatDialogById(String dialogId){
        return dialogsMap.get(dialogId);
    }

    public void clear() {
        dialogsMap.clear();
    }

    public void addDialog(QBChatDialog dialog) {
        if (dialog != null) {
            dialogsMap.put(dialog.getDialogId(), dialog);
        }
    }

    public void addDialogs(List<QBChatDialog> dialogs) {
        for (QBChatDialog dialog : dialogs) {
            addDialog(dialog);
        }
    }

    public void deleteDialogs(Collection<QBChatDialog> dialogs) {
        for (QBChatDialog dialog : dialogs) {
            deleteDialog(dialog);
        }
    }

    public void deleteDialogs(ArrayList<String> dialogsIds) {
        for (String dialogId : dialogsIds) {
            deleteDialog(dialogId);
        }
    }

    public void deleteDialog(QBChatDialog chatDialog){
        dialogsMap.remove(chatDialog.getDialogId());
    }

    public void deleteDialog(String dialogId){
        dialogsMap.remove(dialogId);
    }

    public boolean hasDialogWithId(String dialogId){
        return dialogsMap.containsKey(dialogId);
    }

    public boolean hasPrivateDialogWithUser(QBUser user){
        return getPrivateDialogWithUser(user) != null;
    }

    public QBChatDialog getPrivateDialogWithUser(QBUser user){
        for (QBChatDialog chatDialog : dialogsMap.values()){
            if (QBDialogType.PRIVATE.equals(chatDialog.getType())
                    && chatDialog.getOccupants().contains(user.getId())){
                return chatDialog;
            }
        }

        return null;
    }

    private Map<String, QBChatDialog> getSortedMap(Map <String, QBChatDialog> unsortedMap){
        Map <String, QBChatDialog> sortedMap = new TreeMap(new LastMessageDateSentComparator(unsortedMap));
        sortedMap.putAll(unsortedMap);
        return sortedMap;
    }

    public void updateDialog(String dialogId, QBChatMessage qbChatMessage){
        QBChatDialog updatedDialog = getChatDialogById(dialogId);
        updatedDialog.setLastMessage(qbChatMessage.getBody());
        updatedDialog.setLastMessageDateSent(qbChatMessage.getDateSent());
        updatedDialog.setUnreadMessageCount(updatedDialog.getUnreadMessageCount() != null
                ? updatedDialog.getUnreadMessageCount() + 1 : 1);
        updatedDialog.setLastMessageUserId(qbChatMessage.getSenderId());

        dialogsMap.put(updatedDialog.getDialogId(), updatedDialog);
    }

    static class LastMessageDateSentComparator implements Comparator<String> {
        Map <String, QBChatDialog> map;

        public LastMessageDateSentComparator(Map <String, QBChatDialog> map) {

            this.map = map;
        }

        public int compare(String keyA, String keyB) {

            long valueA = map.get(keyA).getLastMessageDateSent();
            long valueB = map.get(keyB).getLastMessageDateSent();

            if (valueB < valueA){
                return -1;
            } else {
                return 1;
            }
        }
    }
}

File: sample-chat-kotlin/src/main/java/com/quickblox/sample/chat/kotlin/utils/qb/QbDialogHolder.kt
package com.quickblox.sample.chat.kotlin.utils.qb

import com.quickblox.chat.model.QBChatDialog
import com.quickblox.chat.model.QBChatMessage
import com.quickblox.chat.model.QBDialogType
import com.quickblox.users.model.QBUser
import java.util.*

private const val DEFAULT_MESSAGE_COUNT = 1

object QbDialogHolder {

    private val _dialogsMap: MutableMap<String, QBChatDialog> = TreeMap()
    val dialogsMap: Map<String, QBChatDialog>
        get() = getSortedMap(_dialogsMap)

    fun getChatDialogById(dialogId: String): QBChatDialog? {
        return _dialogsMap[dialogId]
    }

    fun clear() {
        _dialogsMap.clear()
    }

    fun addDialog(dialog: QBChatDialog?) {
        dialog?.let {
            _dialogsMap[it.dialogId] = it
        }
    }

    fun addDialogs(dialogs: List<QBChatDialog>) {
        for (dialog in dialogs) {
            addDialog(dialog)
        }
    }

    fun deleteDialogs(dialogsIds: ArrayList<String>) {
        for (dialogId in dialogsIds) {
            deleteDialog(dialogId)
        }
    }

    fun deleteDialog(chatDialog: QBChatDialog) {
        _dialogsMap.remove(chatDialog.dialogId)
    }

    private fun deleteDialog(dialogId: String) {
        _dialogsMap.remove(dialogId)
    }

    fun hasDialogWithId(dialogId: String): Boolean {
        return _dialogsMap.containsKey(dialogId)
    }

    fun hasPrivateDialogWithUser(user: QBUser): Boolean {
        return getPrivateDialogWithUser(user) != null
    }

    fun getPrivateDialogWithUser(user: QBUser): QBChatDialog? {
        var privateDialog: QBChatDialog? = null
        for (chatDialog in _dialogsMap.values) {
            if (QBDialogType.PRIVATE == chatDialog.type && chatDialog.occupants.contains(user.id)) {
                privateDialog = chatDialog
            }
        }
        return privateDialog
    }

    private fun getSortedMap(unsortedMap: Map<String, QBChatDialog>): Map<String, QBChatDialog> {
        val sortedMap = TreeMap<String, QBChatDialog>(LastMessageDateSentComparator(unsortedMap))
        sortedMap.putAll(unsortedMap)
        return sortedMap
    }

    fun updateDialog(dialogId: String, qbChatMessage: QBChatMessage) {
        val updatedDialog = getChatDialogById(dialogId)
        updatedDialog?.let {
            it.lastMessage = qbChatMessage.body
            it.lastMessageDateSent = qbChatMessage.dateSent
            val messageCount = if (it.unreadMessageCount != null) {
                updatedDialog.unreadMessageCount + 1
            } else {
                DEFAULT_MESSAGE_COUNT
            }
            it.unreadMessageCount = messageCount
            it.lastMessageUserId = qbChatMessage.senderId
            _dialogsMap[updatedDialog.dialogId] = updatedDialog
        }
    }

    internal class LastMessageDateSentComparator(var map: Map<String, QBChatDialog>) : Comparator<String> {
        override fun compare(keyA: String, keyB: String): Int {
            val valueA = map[keyA]?.lastMessageDateSent ?: return -1
            val valueB = map[keyB]?.lastMessageDateSent ?: return -1
            val result = if (valueB < valueA) {
                -1
            } else {
                1
            }
            return result
        }
    }
}